stanford-sw/src/edu/stanford/StanfordIndexer.java

```diff
@@ -588,6 +588,64 @@ MSS MEDIA (e.g. MSS MEDIA 0025), MSS PHOTO (e.g. MSS PHOTO 0463), MSS PRINTS (e.
     return physicalFormats;
   }
 
+  public Set<String> getAllGenres(final Record record)
+  {
+    Set<String> freeFormGenres = new HashSet<String>();
+    Set<String> genres = new HashSet<String>();
+    Set<String> allGenres = new HashSet<String>();
+    freeFormGenres = getFreeFormGenre(record, "[\\\\,;:]", "([\\p{L}\\p{N}]{4}|\\.*?[\\s)]|[..{2,}]|[AMUaw][adir][cirt])");
+    genres = getGenres(record);
+    allGenres.addAll(freeFormGenres);
+    allGenres.addAll(genres);
+
+    return allGenres;
+  }
+
+  /**
+   * Returns all 655av and the subfield v (repeatable) in list of 6xx fields
+   * @param record a marc4j Record object
+     * @param charsToReplaceRegEx a regular expression of trailing chars to be
+     *   replaced (see java Pattern class).  Note that the regular expression
+     *   should NOT have '$' at the end.
+     *   (e.g. " *[,/;:]" replaces any commas, slashes, semicolons or colons
+     *     at the end of the string, and these chars may optionally be preceded
+     *     by a space)
+     * @param charsB4periodRegEx a regular expression that must immediately
+     *  precede a trailing period IN ORDER FOR THE PERIOD TO BE REMOVED.
+     *  Note that the regular expression will NOT have the period or '$' at
+     *  the end.
+     *   (e.g. "[a-zA-Z]{3,}" means at least three letters must immediately
+     *   precede the period for it to be removed.)
+   *
+   * @return Set of strings containing genre_ssim values without trailing chars
+   */
+    @SuppressWarnings("unchecked")
+  public Set<String> getFreeFormGenre(final Record record, String charsToReplaceRegEx, String charsB4periodRegEx)
+    {
+    Set<String> values = MarcUtils.getFieldList(record, "655av");
+
+    // look for subfield v (repeatable) in specific 6XX tags
+    String desiredTagFldSpec = "600:610:611:630:647:648:650:651:654:656:657";
+    List<VariableField> list6XXvf = MarcUtils.getVariableFields(record, desiredTagFldSpec);
+    for (VariableField vf : list6XXvf)
+    {
+      DataField df = (DataField) vf;
+      List<String> subList = MarcUtils.getSubfieldStrings(df, 'v');
+      for (int i = 0; i<subList.size(); i++) {
+        values.add(subList.get(i));
```

Comment:
I don't see how to loop through once, since in one loop, I'm adding 6xx subfield(s) v to a list that includes 655av, and in the other loop, I'm normalizing the data.